packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/reducer.js

```diff
@@ -72,29 +73,140 @@ const reducer = (state, action) =>
+
+        if (!connectedRelations) {
+          set(draftState, [...path, 'connect'], []);
         }
 
-        const el = action.value[0].value;
+        if (replace) {
+          set(draftState, [...path, 'connect'], [value]);
+        } else {
+          const nextValue = get(draftState, [...path, 'connect']);
+          nextValue.push(value);
+        }
 
-        const currentValue = get(state, ['modifiedData', ...action.keys], null);
+        if (disconnectedRelations?.length) {
+          const existsInDisconnect = disconnectedRelations.find(
+            (disconnectValue) => disconnectValue.id === value.id
+          );
 
-        if (!currentValue) {
-          set(draftState, ['modifiedData', ...action.keys], [el]);
+          if (existsInDisconnect) {
+            const newDisconnectArray = pull([...disconnectedRelations], existsInDisconnect);
+            set(draftState, [...path, 'disconnect'], newDisconnectArray);
+          }
+        }
 
-          break;
+        break;
+      }
+      case 'DISCONNECT_RELATION': {
+        const path = ['modifiedData', ...action.keys];
+        const { value } = action;
+        const disconnectedRelations = get(state, [...path, 'disconnect']);
+        const connectedRelations = get(state, [...path, 'connect']);
+
+        if (!disconnectedRelations) {
+          set(draftState, [...path, 'disconnect'], []);
         }
 
-        set(draftState, ['modifiedData', ...action.keys], [...currentValue, el]);
+        const nextValue = get(draftState, [...path, 'disconnect']);
+        nextValue.push(value);
+
+        if (connectedRelations?.length) {
+          const existsInConnect = connectedRelations.find(
+            (connectValue) => connectValue.id === value.id
+          );
+
+          if (existsInConnect) {
+            const newConnectArray = pull([...connectedRelations], existsInConnect);
+            set(draftState, [...path, 'connect'], newConnectArray);
+          }
+        }
 
         break;
       }
       case 'INIT_FORM': {
+        const { initialValues, relationalFields = [] } = action;
+
         draftState.formErrors = {};
-        draftState.initialData = action.initialValues;
-        draftState.modifiedData = action.initialValues;
+
+        draftState.initialData = {
+          ...initialValues,
+
+          /**
+           * The state we keep in the client for relations looks like:
+           *
+           * {
+           *   count: <int>
+           *   results: [<Relation>]
+           * }
+           *
+           * The content API only returns { count: <int> }, which is why
+           * we need to extend the existing state rather than overwriting it.
+           */
+
+          ...relationalFields.reduce((acc, name) => {
+            acc[name] = {
+              ...(state.initialData?.[name] ?? {}),
+              ...(initialValues?.[name] ?? {}),
+            };
+
+            return acc;
+          }, {}),
+        };
+
+        draftState.modifiedData = {
```

Comment:
Isn't it a bit breaking change? We have documented that plugins can access the state of the CM

Reply:
Theoretically, it is a breaking change, yes. There is a but: currently, it is broken, and the state of relations is not reflected in the state, which is why I think it is ok to do that. Do you have any other ideas about how to design this backward compatible?